File: mobile/src/main/java/com/thewizrd/simplewear/services/TorchService.java
package com.thewizrd.simplewear.services;

import android.Manifest;
import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.hardware.camera2.CameraManager;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;

import androidx.core.app.NotificationCompat;
import androidx.core.content.ContextCompat;

import com.thewizrd.shared_resources.helpers.Actions;
import com.thewizrd.shared_resources.utils.Logger;
import com.thewizrd.simplewear.App;
import com.thewizrd.simplewear.PhoneBroadcastReceiver;
import com.thewizrd.simplewear.R;
import com.thewizrd.simplewear.wearable.WearableWorker;

public class TorchService extends Service {

    private static final int JOB_ID = 1001;
    private static final String NOT_CHANNEL_ID = "SimpleWear.torchservice";

    public static final String ACTION_START_LIGHT = "SimpleWear.Droid.action.START_LIGHT";
    public static final String ACTION_END_LIGHT = "SimpleWear.Droid.action.END_LIGHT";

    private boolean mFlashEnabled = false;

    public static void enqueueWork(Context context, Intent work) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(work);
        } else {
            context.startService(work);
        }
    }

    private static void initChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // Gets an instance of the NotificationManager service
            Context context = App.getInstance().getAppContext();
            NotificationManager mNotifyMgr = (NotificationManager) context.getSystemService(NOTIFICATION_SERVICE);
            NotificationChannel mChannel = mNotifyMgr.getNotificationChannel(NOT_CHANNEL_ID);

            if (mChannel == null) {
                String notchannel_name = context.getString(R.string.not_channel_name_torch);

                mChannel = new NotificationChannel(NOT_CHANNEL_ID, notchannel_name, NotificationManager.IMPORTANCE_LOW);
                // Configure the notification channel.
                mChannel.setShowBadge(false);
                mChannel.enableLights(false);
                mChannel.enableVibration(false);
                mNotifyMgr.createNotificationChannel(mChannel);
            }
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private static Notification getForegroundNotification() {
        Context context = App.getInstance().getAppContext();
        NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(context, NOT_CHANNEL_ID)
                        .setSmallIcon(R.drawable.ic_lightbulb_outline_white_24dp)
                        .setContentTitle(context.getString(R.string.action_torch))
                        .setContentText(context.getString(R.string.not_torch_turnoff_summary))
                        .setContentIntent(PendingIntent.getBroadcast(context, JOB_ID,
                                new Intent(context, PhoneBroadcastReceiver.class).setAction(ACTION_END_LIGHT), PendingIntent.FLAG_UPDATE_CURRENT))
                        .setColor(ContextCompat.getColor(context, R.color.colorPrimary))
                        .setOnlyAlertOnce(true)
                        .setPriority(NotificationCompat.PRIORITY_HIGH);

        return mBuilder.build();
    }

    @Override
    public void onCreate() {
        super.onCreate();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            initChannel();
        }

        startForeground(JOB_ID, getForegroundNotification());
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        startForeground(JOB_ID, getForegroundNotification());

        if (intent != null && ACTION_START_LIGHT.equals(intent.getAction())) {
            turnOnFlashLight();
        } else if (intent != null && ACTION_END_LIGHT.equals(intent.getAction())) {
            turnOffFlashLight();
            stopSelf();
        }

        WearableWorker.sendActionUpdate(getApplicationContext(), Actions.TORCH);

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        if (mFlashEnabled) {
            turnOffFlashLight();
        }

        stopForeground(true);

        super.onDestroy();
    }

    private void turnOnFlashLight() {
        final boolean hasCameraFlash = getPackageManager().
                hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH);
        boolean isEnabled = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_GRANTED;

        if (!hasCameraFlash || !isEnabled) {
            stopSelf();
            return;
        }

        try {
            CameraManager cameraManager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
            String cameraId = cameraManager.getCameraIdList()[0];
            cameraManager.setTorchMode(cameraId, true);
            mFlashEnabled = true;
        } catch (Exception e) {
            Logger.writeLine(Log.ERROR, e);
        }
    }

    private void turnOffFlashLight() {
        final boolean hasCameraFlash = getPackageManager().
                hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH);
        boolean isEnabled = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_GRANTED;

        if (!hasCameraFlash || !isEnabled) {
            stopSelf();
            return;
        }

        CameraManager cameraManager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
        String cameraId = null;
        try {
            cameraId = cameraManager.getCameraIdList()[0];
            cameraManager.setTorchMode(cameraId, false);
            mFlashEnabled = false;
        } catch (Exception e) {
            Logger.writeLine(Log.ERROR, e);
        }
    }
}

File: mobile/src/main/java/com/thewizrd/simplewear/services/TorchService.kt
package com.thewizrd.simplewear.services

import android.Manifest
import android.app.*
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.hardware.camera2.CameraManager
import android.os.Build
import android.os.IBinder
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import com.thewizrd.shared_resources.actions.Actions
import com.thewizrd.shared_resources.helpers.toImmutableCompatFlag
import com.thewizrd.shared_resources.utils.Logger
import com.thewizrd.simplewear.R
import com.thewizrd.simplewear.receivers.PhoneBroadcastReceiver
import com.thewizrd.simplewear.wearable.WearableWorker

class TorchService : Service() {
    companion object {
        private const val JOB_ID = 1001
        private const val NOT_CHANNEL_ID = "SimpleWear.torchservice"

        const val ACTION_START_LIGHT = "SimpleWear.Droid.action.START_LIGHT"
        const val ACTION_END_LIGHT = "SimpleWear.Droid.action.END_LIGHT"

        fun enqueueWork(context: Context, work: Intent) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(work)
            } else {
                context.startService(work)
            }
        }
    }

    private var mFlashEnabled = false

    override fun onCreate() {
        super.onCreate()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            initChannel()
        }
        startForeground(JOB_ID, getForegroundNotification(applicationContext))
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    @RequiresApi(Build.VERSION_CODES.O)
    private fun initChannel() {
        // Gets an instance of the NotificationManager service
        val mNotifyMgr =
            applicationContext.getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        var mChannel = mNotifyMgr.getNotificationChannel(NOT_CHANNEL_ID)
        val notChannelName = applicationContext.getString(R.string.not_channel_name_torch)
        if (mChannel == null) {
            mChannel = NotificationChannel(
                NOT_CHANNEL_ID,
                notChannelName,
                NotificationManager.IMPORTANCE_DEFAULT
            )
        }

        // Configure the notification channel.
        mChannel.name = notChannelName
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            if (!mChannel.hasUserSetImportance()) {
                mChannel.importance = NotificationManager.IMPORTANCE_DEFAULT
            }
        }
        mChannel.setShowBadge(false)
        mChannel.enableLights(false)
        mChannel.enableVibration(false)
        mNotifyMgr.createNotificationChannel(mChannel)
    }

    private fun getForegroundNotification(context: Context): Notification {
        val mBuilder = NotificationCompat.Builder(context, NOT_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_lightbulb_outline_white_24dp)
            .setContentTitle(context.getString(R.string.action_torch))
            .addAction(
                0,
                context.getString(R.string.action_turnoff),
                PendingIntent.getBroadcast(
                    context, JOB_ID,
                    Intent(context, PhoneBroadcastReceiver::class.java)
                        .setAction(ACTION_END_LIGHT),
                    PendingIntent.FLAG_UPDATE_CURRENT.toImmutableCompatFlag()
                )
            )
            .setColor(ContextCompat.getColor(context, R.color.colorPrimary))
            .setOnlyAlertOnce(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
        return mBuilder.build()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        startForeground(JOB_ID, getForegroundNotification(applicationContext))

        when (intent?.action) {
            ACTION_START_LIGHT -> {
                turnOnFlashLight()
            }
            ACTION_END_LIGHT -> {
                turnOffFlashLight()
                stopSelf()
            }
        }

        WearableWorker.sendActionUpdate(applicationContext, Actions.TORCH)

        return START_STICKY
    }

    override fun onDestroy() {
        if (mFlashEnabled) {
            turnOffFlashLight()
        }
        stopForeground(true)
        super.onDestroy()
    }

    private fun turnOnFlashLight() {
        val hasCameraFlash = packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH)
        val isEnabled = (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED)

        if (!hasCameraFlash || !isEnabled) {
            stopSelf()
            return
        }

        try {
            val cameraManager = getSystemService(CAMERA_SERVICE) as CameraManager
            val cameraId = cameraManager.cameraIdList[0]
            cameraManager.setTorchMode(cameraId, true)
            mFlashEnabled = true
        } catch (e: Exception) {
            Logger.writeLine(Log.ERROR, e)
        }
    }

    private fun turnOffFlashLight() {
        val hasCameraFlash = packageManager.hasSystemFeature(PackageManager.FEATURE_CAMERA_FLASH)
        val isEnabled = (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED)

        if (!hasCameraFlash || !isEnabled) {
            stopSelf()
            return
        }

        val cameraManager = getSystemService(CAMERA_SERVICE) as CameraManager
        var cameraId: String? = null

        try {
            cameraId = cameraManager.cameraIdList[0]
            cameraManager.setTorchMode(cameraId, false)
            mFlashEnabled = false
        } catch (e: Exception) {
            Logger.writeLine(Log.ERROR, e)
        }
    }
}